File: spacetimedb/src/systems/movement.ts
import { MOVE_DT, ARRIVE_EPS } from '../../../shared/constants.ts';
import { stepToward } from '../../../shared/sim.ts';
import { spacetimedb } from '../schema/db.ts';
import { moveTimer } from '../schema/tables.ts';
import { scheduleRefs } from '../schema/schedule_refs.ts';
import { activeMatchIds } from '../world/scope.ts';

// Movement integration — runs every MOVE_TICK_MS. Only touches movers in Active
// matches (a Paused/Ended match's units freeze in place).
export const moveUnits = spacetimedb.reducer(
  { timer: moveTimer.rowType },
  (ctx) => {
    const active = activeMatchIds(ctx);
    for (const u of [...ctx.db.unit.iter()]) {
      if (!active.has(u.matchId)) continue; // paused/ended match — frozen
      if (u.garrisonedIn !== 0n) continue; // sheltered — off the field
      if (!u.hasTarget) continue;
      const e = ctx.db.entity.entityId.find(u.entityId);
      if (!e) continue;

      const r = stepToward(
        e.x,
        e.y,
        u.targetX,
        u.targetY,
        u.speed * MOVE_DT,
        ARRIVE_EPS
      );
      ctx.db.entity.entityId.update({ ...e, x: r.x, y: r.y, facing: r.facing });
      if (!r.arrived) continue;
      const next = u.pathIdx + 1;
      if (next < u.path.length) {
        const wp = u.path[next];
        ctx.db.unit.entityId.update({
          ...u,
          pathIdx: next,
          targetX: wp.x,
          targetY: wp.y,
        });
      } else {
        ctx.db.unit.entityId.update({ ...u, hasTarget: false });
      }
    }
  }
);

scheduleRefs.moveUnits = moveUnits;

File: saladin-bevy/crates/protocol/src/systems/movement.rs
use crate::MatchStatuses;
use crate::components::{MatchId, Pos, Unit};
use bevy_ecs::prelude::*;
use saladin_sim::{ARRIVE_EPS, MOVE_DT, step_toward};

/// Integrate every active mover one base tick toward its target, advancing along
/// its path on arrival. Garrisoned units are off the field; paused matches
/// freeze in place. Ported from the SpacetimeDB `moveUnits` reducer; the
/// spatial-grid `cell` is maintained by the index system, not here.
pub fn movement(statuses: Res<MatchStatuses>, mut q: Query<(&mut Pos, &mut Unit, &MatchId)>) {
    for (mut pos, mut u, mid) in &mut q {
        if u.garrisoned_in != 0 || !u.has_target || !statuses.simulates(mid.0) {
            continue;
        }
        let step = u.speed * MOVE_DT;
        let r = step_toward(pos.pos, u.target, step, ARRIVE_EPS);
        // facing is a render hint; derive it from the heading without trig later
        // on the client. Sim stores position only.
        pos.pos = r.pos;
        if !r.arrived {
            continue;
        }
        let next = u.path_idx + 1;
        if next < u.path.len() {
            u.path_idx = next;
            u.target = u.path[next];
        } else {
            u.has_target = false;
        }
    }
}
